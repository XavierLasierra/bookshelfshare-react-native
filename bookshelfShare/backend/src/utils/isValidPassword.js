const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
