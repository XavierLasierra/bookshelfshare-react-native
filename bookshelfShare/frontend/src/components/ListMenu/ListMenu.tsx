import React, {useEffect, useState} from 'react';
import {
  Menu,
  MenuOptions,
  MenuTrigger,
  MenuOption,
  renderers,
} from 'react-native-popup-menu';
import {View, Text, ActivityIndicator} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {IListMenuProps, IStore} from '../../types/interfaces';

import {updateUserBooks} from '../../redux/actions/loggedUser.creator';

import stylesConstants from '../../styles/styles.constants';
import styles from './listMenu.styles';
import globalStyles from '../../styles/global.styles';

export default function ListMenu({
  bookIsbn,
  userId,
  token,
  refreshToken,
}: IListMenuProps) {
  const dispatch = useDispatch();
  const books = useSelector((store: IStore) => store.userBooks);

  const [isLoading, setIsLoading] = useState(false);
  const [markListName, setMarkListName] = useState('Mark as');
  const [deleteFrom, setDeleteFrom] = useState('');

  function findInUserBooks() {
    if (books.read.includes(bookIsbn)) {
      setDeleteFrom('read');
      return 'Read';
    }
    if (books.reading.includes(bookIsbn)) {
      setDeleteFrom('reading');
      return 'Reading';
    }
    if (books.wishlist.includes(bookIsbn)) {
      setDeleteFrom('wishlist');
      return 'Wishlist';
    }
    if (books.toRead.includes(bookIsbn)) {
      setDeleteFrom('toRead');
      return 'To Read';
    }
    return 'Mark as';
  }

  useEffect(() => {
    setMarkListName(findInUserBooks());
    setIsLoading(false);
  }, [books]);

  function handleCategorySelect(value: string) {
    setIsLoading(true);
    dispatch(
      updateUserBooks(
        userId,
        {
          bookIsbn,
          deleteFrom,
          addTo: value,
        },
        token,
        refreshToken,
      ),
    );
  }

  return (
    <Menu
      onSelect={handleCategorySelect}
      renderer={renderers.NotAnimatedContextMenu}>
      <MenuTrigger>
        {isLoading ? (
          <View style={styles.activityIndicatorContainer}>
            <ActivityIndicator
              size="small"
              color={stylesConstants.colors.dark}
            />
          </View>
        ) : (
          <Text style={globalStyles.markButton}>{markListName}</Text>
        )}
      </MenuTrigger>
      <MenuOptions>
        <MenuOption value="read" text="Read" style={globalStyles.menuOption} />
        <MenuOption
          value="reading"
          text="Reading"
          style={globalStyles.menuOption}
        />
        <MenuOption
          value="wishlist"
          text="Wishlist"
          style={globalStyles.menuOption}
        />
        <MenuOption
          value="toRead"
          text="To Read"
          style={globalStyles.menuOption}
        />
        <MenuOption value="" text="Unmark" style={globalStyles.menuOption} />
      </MenuOptions>
    </Menu>
  );
}
