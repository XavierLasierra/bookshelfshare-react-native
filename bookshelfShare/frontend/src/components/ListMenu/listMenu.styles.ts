import {StyleSheet} from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  activityIndicatorContainer: {
    backgroundColor: stylesConstants.colors.white,
    borderRadius: 10,
    borderWidth: 1,
    borderColor: stylesConstants.colors.secondary,
    height: 40,
    width: 150,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
