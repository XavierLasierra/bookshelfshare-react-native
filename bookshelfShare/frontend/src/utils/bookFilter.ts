export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
