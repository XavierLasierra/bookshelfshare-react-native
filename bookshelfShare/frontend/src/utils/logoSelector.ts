import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
