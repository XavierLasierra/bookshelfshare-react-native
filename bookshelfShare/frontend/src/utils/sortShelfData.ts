export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
