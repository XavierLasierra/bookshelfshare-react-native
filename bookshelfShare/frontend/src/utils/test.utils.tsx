import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
