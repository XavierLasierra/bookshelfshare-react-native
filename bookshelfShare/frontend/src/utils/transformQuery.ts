interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
