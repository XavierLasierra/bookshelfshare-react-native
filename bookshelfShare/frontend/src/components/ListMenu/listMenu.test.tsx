import React from 'react';
import {MenuProvider} from 'react-native-popup-menu';
import ListMenu from './ListMenu';
import {render} from '../../utils/test.utils';
import userBooksMock from '../../mocks/userBooks.mock';

const initialState = {userBooks: userBooksMock};

describe('Given a ListMenu component', () => {
  describe('When it is rendered', () => {
    describe('And userBooks does not include bookIsbn', () => {
      test('Then screen should match the snapshot', () => {
        const screen = render(
          <MenuProvider>
            <ListMenu
              bookIsbn="0"
              userId="1"
              token="token"
              refreshToken="refreshToken"
            />
          </MenuProvider>,
          initialState,
        );
        expect(screen).toMatchSnapshot();
      });
    });

    [
      {
        listName: 'read',
        bookIsbn: '1',
      },
      {
        listName: 'reading',
        bookIsbn: '2',
      },
      {
        listName: 'toRead',
        bookIsbn: '3',
      },
      {
        listName: 'wishlist',
        bookIsbn: '4',
      },
    ].forEach(({listName, bookIsbn}: any) => {
      describe(`And userBooks does include bookIsbn in ${listName}`, () => {
        test('Then screen should match the snapshot', () => {
          const screen = render(
            <MenuProvider>
              <ListMenu
                bookIsbn={bookIsbn}
                userId="1"
                token="token"
                refreshToken="refreshToken"
              />
            </MenuProvider>,
            initialState,
          );
          expect(screen).toMatchSnapshot();
        });
      });
    });
  });
});
