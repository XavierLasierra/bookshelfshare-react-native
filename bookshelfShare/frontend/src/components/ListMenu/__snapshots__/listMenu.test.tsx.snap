// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ListMenu component When it is rendered And userBooks does include bookIsbn in read Then screen should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            ctx={
              Object {
                "menuActions": Object {
                  "_getOpenedMenu": [Function],
                  "_notify": [Function],
                  "closeMenu": [Function],
                  "isMenuOpen": [Function],
                  "openMenu": [Function],
                  "toggleMenu": [Function],
                },
                "menuRegistry": Object {
                  "getAll": [Function],
                  "getMenu": [Function],
                  "setOptionsCustomStyles": [Function],
                  "subscribe": [Function],
                  "unsubscribe": [Function],
                  "updateLayoutInfo": [Function],
                },
              }
            }
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#ECECEC",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 40,
                  "lineHeight": 40,
                  "textAlign": "center",
                  "width": 150,
                }
              }
            >
              Read
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;

exports[`Given a ListMenu component When it is rendered And userBooks does include bookIsbn in reading Then screen should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            ctx={
              Object {
                "menuActions": Object {
                  "_getOpenedMenu": [Function],
                  "_notify": [Function],
                  "closeMenu": [Function],
                  "isMenuOpen": [Function],
                  "openMenu": [Function],
                  "toggleMenu": [Function],
                },
                "menuRegistry": Object {
                  "getAll": [Function],
                  "getMenu": [Function],
                  "setOptionsCustomStyles": [Function],
                  "subscribe": [Function],
                  "unsubscribe": [Function],
                  "updateLayoutInfo": [Function],
                },
              }
            }
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#ECECEC",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 40,
                  "lineHeight": 40,
                  "textAlign": "center",
                  "width": 150,
                }
              }
            >
              Reading
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;

exports[`Given a ListMenu component When it is rendered And userBooks does include bookIsbn in toRead Then screen should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            ctx={
              Object {
                "menuActions": Object {
                  "_getOpenedMenu": [Function],
                  "_notify": [Function],
                  "closeMenu": [Function],
                  "isMenuOpen": [Function],
                  "openMenu": [Function],
                  "toggleMenu": [Function],
                },
                "menuRegistry": Object {
                  "getAll": [Function],
                  "getMenu": [Function],
                  "setOptionsCustomStyles": [Function],
                  "subscribe": [Function],
                  "unsubscribe": [Function],
                  "updateLayoutInfo": [Function],
                },
              }
            }
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#ECECEC",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 40,
                  "lineHeight": 40,
                  "textAlign": "center",
                  "width": 150,
                }
              }
            >
              To Read
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;

exports[`Given a ListMenu component When it is rendered And userBooks does include bookIsbn in wishlist Then screen should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            ctx={
              Object {
                "menuActions": Object {
                  "_getOpenedMenu": [Function],
                  "_notify": [Function],
                  "closeMenu": [Function],
                  "isMenuOpen": [Function],
                  "openMenu": [Function],
                  "toggleMenu": [Function],
                },
                "menuRegistry": Object {
                  "getAll": [Function],
                  "getMenu": [Function],
                  "setOptionsCustomStyles": [Function],
                  "subscribe": [Function],
                  "unsubscribe": [Function],
                  "updateLayoutInfo": [Function],
                },
              }
            }
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#ECECEC",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 40,
                  "lineHeight": 40,
                  "textAlign": "center",
                  "width": 150,
                }
              }
            >
              Wishlist
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;

exports[`Given a ListMenu component When it is rendered And userBooks does not include bookIsbn Then screen should match the snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View>
      <View
        collapsable={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            ctx={
              Object {
                "menuActions": Object {
                  "_getOpenedMenu": [Function],
                  "_notify": [Function],
                  "closeMenu": [Function],
                  "isMenuOpen": [Function],
                  "openMenu": [Function],
                  "toggleMenu": [Function],
                },
                "menuRegistry": Object {
                  "getAll": [Function],
                  "getMenu": [Function],
                  "setOptionsCustomStyles": [Function],
                  "subscribe": [Function],
                  "unsubscribe": [Function],
                  "updateLayoutInfo": [Function],
                },
              }
            }
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#ECECEC",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "height": 40,
                  "lineHeight": 40,
                  "textAlign": "center",
                  "width": 150,
                }
              }
            >
              Mark as
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </RCTSafeAreaView>
</View>
`;
